peekingduck/viewer/viewer.py

```diff
@@ -0,0 +1,583 @@
+# Copyright 2022 AI Singapore
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Implement PeekingDuck Viewer
+"""
+
+from typing import List
+from pathlib import Path
+import logging
+import platform
+import tkinter as tk
+from tkinter import filedialog
+from tkinter.messagebox import askyesno
+import threading
+import copy
+import cv2
+import numpy as np
+from PIL import Image, ImageTk
+from peekingduck.declarative_loader import DeclarativeLoader
+from peekingduck.pipeline.pipeline import Pipeline
+from peekingduck.viewer.playlist import PlayList
+from peekingduck.viewer.viewer_gui import create_window
+from peekingduck.viewer.viewer_utils import (
+    get_keyboard_char,
+    get_keyboard_modifier,
+)
+
+####################
+# Globals
+####################
+BUTTON_DELAY: int = 250  # milliseconds (0.25 of a second)
+BUTTON_REPEAT: int = int(1000 / 60)  # milliseconds (60 fps)
+CHANGE_PIPELINE_DELAY: float = 1.0  # seconds before switching pipelines
+FPS_60: int = int(1000 / 60)  # milliseconds per iteration
+ZOOM_TEXT: List[str] = ["50%", "75%", "100%", "125%", "150%", "200%", "250%", "300%"]
+ZOOM_DEFAULT_IDX: int = 2
+ZOOMS: List[float] = [0.5, 0.75, 1.0, 1.25, 1.50, 2.00, 2.50, 3.00]  # > 3x is slow!
+EMOJI_PLAY = "\u25B6"
+EMOJI_STOP = "\u25FD"  # white medium-small square
+
+
+class Viewer:  # pylint: disable=too-many-instance-attributes, too-many-public-methods
+    """Implement PeekingDuck Viewer class"""
+
+    def __init__(
+        self,
+        pipeline_path: Path,
+        config_updates_cli: str,
+        custom_nodes_parent_subdir: str,
+        num_iter: int = 0,
+    ) -> None:
+        self.logger = logging.getLogger(__name__)
```

Comment:
The viewer has introduced a large number of logging messages which are probably not required by the typical user. Suggest moving most of them to `DEBUG` level and keeping only a few such as `quitting viewer` or `XXX.yaml already in playlist` at `INFO` level.

```
(makerpose) Ngs-MacBook-Pro:PeekingDuck ngleeping$ peekingduck run --viewer
2022-07-06 13:16:47 peekingduck.cli  INFO:  Launching PeekingDuck Viewer 
2022-07-06 13:16:47 peekingduck.viewer.playlist  INFO:  playlist_path=/Users/ngleeping/.peekingduck/playlist.yaml 
2022-07-06 13:16:47 peekingduck.viewer.viewer  INFO:  init pipeline: pipeline_path=/Users/ngleeping/Repositories/PeekingDuck/pipeline_config.yml, is_abs=True 
2022-07-06 13:16:47 peekingduck.viewer.viewer  INFO:  custom nodes parent: <class 'str'> 
2022-07-06 13:16:47 peekingduck.viewer.playlist  INFO:  /Users/ngleeping/Repositories/PeekingDuck/pipeline_config.yml already in playlist 
2022-07-06 13:16:47 peekingduck.viewer.viewer  INFO:  cwd=/Users/ngleeping/Repositories/PeekingDuck 
2022-07-06 13:16:47 peekingduck.viewer.viewer  INFO:  pipeline=/Users/ngleeping/Repositories/PeekingDuck/pipeline_config.yml 
2022-07-06 13:16:47 peekingduck.viewer.viewer  INFO:  binding macOS cmd-Q 
2022-07-06 13:16:47 peekingduck.declarative_loader  INFO:  Successfully loaded pipeline file. 
2022-07-06 13:16:47 peekingduck.declarative_loader  INFO:  Initializing input.visual node... 
2022-07-06 13:16:47 peekingduck.pipeline.nodes.input.visual  INFO:  Input size: 710 by 540 
2022-07-06 13:16:47 peekingduck.declarative_loader  INFO:  Initializing model.yolo node... 
2022-07-06 13:16:52 peekingduck.pipeline.nodes.model.yolov4.yolo_files.detector  INFO:  YOLO model loaded with following configs: 
	Model type: v4tiny, 
	Input resolution: (416, 416), 
	IDs being detected: [0] 
	Max detections per class: 50, 
	Max total detections: 50, 
	IOU threshold: 0.5, 
	Score threshold: 0.2 
2022-07-06 13:16:53 peekingduck.declarative_loader  INFO:  Initializing draw.bbox node... 
2022-07-06 13:16:53 peekingduck.declarative_loader  INFO:  Initializing output.screen node... 
2022-07-06 13:16:57 peekingduck.viewer.single_column_view  INFO:  show stats[2]: pipeline_config.yml 
2022-07-06 13:17:04 peekingduck.pipeline.nodes.input.visual  INFO:  Approximate Progress: 10% 
2022-07-06 13:17:06 peekingduck.viewer.single_column_view  INFO:  show stats[0]: pipeline_config.yml 
2022-07-06 13:17:06 peekingduck.viewer.single_column_view  INFO:  show stats[2]: pipeline_config.yml 
2022-07-06 13:17:07 peekingduck.viewer.single_column_view  INFO:  show stats[0]: pipeline_config.yml 
2022-07-06 13:17:07 peekingduck.viewer.single_column_view  INFO:  show stats[2]: pipeline_config.yml 
2022-07-06 13:17:08 peekingduck.viewer.single_column_view  INFO:  show stats[0]: pipeline_config.yml 
2022-07-06 13:17:08 peekingduck.viewer.single_column_view  INFO:  show stats[2]: pipeline_config.yml 
2022-07-06 13:17:09 peekingduck.pipeline.nodes.input.visual  INFO:  Frames Processed: 100 
2022-07-06 13:17:09 peekingduck.viewer.single_column_view  INFO:  show stats[0]: pipeline_config.yml 
2022-07-06 13:17:10 peekingduck.pipeline.nodes.input.visual  INFO:  Approximate Progress: 20% 
2022-07-06 13:17:10 peekingduck.viewer.single_column_view  INFO:  show stats[2]: pipeline_config.yml 
2022-07-06 13:17:12 peekingduck.viewer.playlist  INFO:  playlist_dict={'playlist': ['/Users/ngleeping/Repositories/PeekingDuck/pipeline_config.yml', '/Users/ngleeping/Desktop/pipeline_config.yml', '/Users/ngleeping/Desktop/pipeline_config 2.yml']} 
2022-07-06 13:17:12 peekingduck.viewer.viewer  INFO:  quitting viewer
```

Reply:
Will remove redundant development debugging outputs, and change the others to `debug` instead of `info`.